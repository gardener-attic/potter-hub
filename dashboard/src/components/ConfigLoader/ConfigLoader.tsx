import * as React from "react";

import UnexpectedErrorPage from "../../components/ErrorAlert/UnexpectedErrorAlert";
import LoadingWrapper, { ILoadingWrapperProps } from "../../components/LoadingWrapper";

interface IConfigLoaderProps extends ILoadingWrapperProps {
  getConfig: () => void;
  error?: Error;
}

class ConfigLoader extends React.Component<IConfigLoaderProps> {
  public componentDidMount() {
    this.props.getConfig();
  }

  public render() {
    const { error, ...otherProps } = this.props;
    if (error) {
      return (
        <UnexpectedErrorPage
          raw={true}
          showGenericMessage={true}
          text={`Unable to load app hub configuration: ${error.message}`}
        />
      );
    }
    return <LoadingWrapper {...otherProps} />;
  }
}

export default ConfigLoader;
