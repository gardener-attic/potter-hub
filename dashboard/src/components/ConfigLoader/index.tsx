import ConfigLoader from "./ConfigLoader";

export default ConfigLoader;
