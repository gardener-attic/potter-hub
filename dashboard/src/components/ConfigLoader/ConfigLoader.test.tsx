import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import ConfigLoader from ".";
import UnexpectedErrorPage from "../../components/ErrorAlert/UnexpectedErrorAlert";
import itBehavesLike from "../../shared/specs";

context("when the config is not ready", () => {
  itBehavesLike("aLoadingComponent", {
    component: ConfigLoader,
    props: {
      loaded: false,
      getConfig: jest.fn(),
    },
  });
});

context("when there is an error", () => {
  it("renders the error details", () => {
    const wrapper = shallow(
      <ConfigLoader error={new Error("Wrong config!")} getConfig={jest.fn()} />,
    );
    expect(wrapper.find(UnexpectedErrorPage)).toExist();
    expect(wrapper).toMatchSnapshot();
  });
});

describe("componentDidMount", () => {
  it("calls getConfig", () => {
    const getConfig = jest.fn();
    shallow(<ConfigLoader getConfig={getConfig} />);
    expect(getConfig).toHaveBeenCalled();
  });
});
