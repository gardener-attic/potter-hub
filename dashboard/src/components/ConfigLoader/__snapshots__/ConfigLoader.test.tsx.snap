// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the config is not ready loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  getConfig={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  loaded={false}
  type={0}
/>
`;

exports[`when there is an error renders the error details 1`] = `
<UnexpectedErrorPage
  raw={true}
  showGenericMessage={true}
  text="Unable to load app hub configuration: Wrong config!"
  title="Sorry! Something went wrong."
/>
`;
